### PR TITLE
Make `i*::signum` a `const fn`.

### PR DESCRIPTION
**Pull request by ecstatic-morse**
Ticks a box in #53718.

This uses a well-known branchless implementation of `signum`: `(n > 0) as i32 - (n < 0) as i32`. 

Here's a [playground](https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=747cf191c4974bf66c9d75e509ae6e6e) comparing the two techniques. On x86 in release mode, the branchless implementation is able to replace a `mov` and `cmov` with a `sar` and `add`, so this should be a bit faster as well.

~~This is marked as a draft since I think I'll need to add `#[rustc_const_unstable]` somewhere. Perhaps the reviewer can point me in the right direction.~~

### CODE REVIEW

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -1994,12 +1994,8 @@ $EndFeature, "
 ```"),
             #[stable(feature = "rust1", since = "1.0.0")]
             #[inline]
-            pub fn signum(self) -> Self {
-                match self {
-                    n if n > 0 =>  1,
-                    0          =>  0,
-                    _          => -1,
-                }
+            pub const fn signum(self) -> Self {
```
Comment (oli-obk) [inline src/libcore/num/mod.rs]: You can just invent new feature gate names with `rustc_const_unstable`, so the following should work

```suggestion
            #[rustc_const_unstable(feature = "const_signum")]
            pub const fn signum(self) -> Self {
```

**Review by oli-obk — CHANGES_REQUESTED** [review]
r? @oli-obk

File: src/libcore/num/mod.rs (hunk 1 above)
Comment (ecstatic-morse) [reply src/libcore/num/mod.rs]: Okay, wasn't sure if I had to define it somewhere else. On #53718 it's under the  "const_int_sign" heading so I think I'll just use that?

Comment (oli-obk) [reply src/libcore/num/mod.rs]: sounds good to me

**Comment by oli-obk** [comment]
can you also add some tests using this inside a constant and comparing the result with expected values?

There should already be tests for other const fns for integer types, you can add the new test to them.

File: src/test/run-pass/const-int-sign.rs [hunk 2]
```diff
@@ -1,11 +1,19 @@
 const NEGATIVE_A: bool = (-10i32).is_negative();
```
Comment (oli-obk) [inline src/test/run-pass/const-int-sign.rs]: I think these test files now need `#![feature(const_int_sign)]`

File: src/test/ui/consts/const-int-sign.rs [hunk 3]
```diff
@@ -3,4 +3,7 @@ fn main() {
     //~^ ERROR temporary value dropped while borrowed
     let y: &'static bool = &(5_i32.is_positive());
     //~^ ERROR temporary value dropped while borrowed
+
+    let sgn: &'static i32 = &(5_i32.signum());
+    //~^ ERROR temporary value dropped while borrowed
```
Comment (ecstatic-morse) [inline src/test/ui/consts/const-int-sign.rs]: I think this file is supposed to test that results of calls to const fns aren't promoted? Do I need to add to this?

File: src/test/run-pass/const-int-sign.rs [hunk 4]
```diff
@@ -1,11 +1,19 @@
...
 const NEGATIVE_B: bool = 10i32.is_negative();
-const POSITIVE_A: bool= (-10i32).is_positive();
-const POSITIVE_B: bool= 10i32.is_positive();
+const POSITIVE_A: bool = (-10i32).is_positive();
+const POSITIVE_B: bool = 10i32.is_positive();
```
Comment (ecstatic-morse) [inline src/test/run-pass/const-int-sign.rs]: Is it okay to include this change here?

File: src/test/run-pass/const-int-sign.rs (hunk 2 above)
Comment (ecstatic-morse) [reply src/test/run-pass/const-int-sign.rs]: Indeed. Sorry, I was relying on CI to run the tests.

File: src/test/run-pass/const-int-sign.rs (hunk 4 above)
Comment (oli-obk) [reply src/test/run-pass/const-int-sign.rs]: jup

File: src/test/ui/consts/const-int-sign.rs (hunk 3 above)
Comment (oli-obk) [reply src/test/ui/consts/const-int-sign.rs]: not really, you can remove it again if you want

Comment (ecstatic-morse) [reply src/test/ui/consts/const-int-sign.rs]: They're gone now. I'll ping you when this passes CI.

**Comment by ecstatic-morse** [comment]
@oli-obk I think this is ready to go now.

**Comment by bjacob** [comment]
For what it's worth, regarding 'branchless' in the commit message,  this code was/is branchless and more or less as efficient before and after this commit just the same.  Example on ARM64:
https://godbolt.org/z/9w4BPr
The old form compiled to 2 mask-and-select instructions (0 branches).
The new form compiles to 1 mask-and-select and 1 subtract-with-shifted-operand (it shifts by 31 bits to get the sign bit, that's how it evals (x < 0)).
So it's just 2 arithmetic instructions before and after.

Note though that the old form also informed the compiler that the result was either -1, 0 or +1.  The new form looks to the compiler like the result might be any integer value.  That might have side effects on optimization opportunities in inlined contexts.

**Comment by ecstatic-morse** [comment]
@bjacob, I compared the generated assembly (albeit only on x86) in the [PR message](https://github.com/rust-lang/rust/pull/61635#issue-286337199), so I'm aware that the optimized version of the old implementation is branchless. I've reproduced the full output below for you.
```asm
playground::signum: # @playground::signum
# %bb.0:
	xorl	%eax, %eax
	testl	%edi, %edi
	setg	%al
	sarl	$31, %edi
	addl	%edi, %eax
	retq
                                        # -- End function

playground::signum_branch: # @playground::signum_branch
# %bb.0:
	xorl	%ecx, %ecx
	testl	%edi, %edi
	setne	%cl
	movl	$-1, %eax
	cmovnsl	%ecx, %eax
	retq
                                        # -- End function
```

The main benefit of this PR is that `signum` is now callable in a `const` context (since the implementation has no branches). A side benefit is that the code is [now faster](http://quick-bench.com/WQ-CKJkXaDAdEGEhqH_jl0p5X8s) (caveats: on x86, using `clang` instead of `rustc`, in a contrived microbenchmark, on whatever machine quick-bench.com uses). `cmov` is slower than your typical, constant-time arithmetic instruction on x86 (although the gap is closing with each new microarchitecture), and the new version uses one less register (on both ARM and x86).

Your last point is an interesting one, although I'm not aware of a set of optimizations that would make use this information. Perhaps you have an idea?

**Comment by bjacob** [comment]
right, sorry I had not followed the link in the commit message.
about the latter point, I was thinking about this: https://godbolt.org/z/kfha8c
See how switch_foo (old form) is simpler than switch_bar(new form) just because the logic is transparent to the compiler so it's able to merge the inline code at the logical level and not even evaluate the actual 'signum' value.

**Comment by bjacob** [comment]
On the other hand, I'm impressed that even in the new form, the compiler was able to deduce that the result could only be -1, 0 or 1 so it still elided the default: case.

**Comment by ecstatic-morse** [comment]
Ah, I didn't consider branching on the result of `signum`. The original implementation will indeed be faster in that case since the compiler can unify the control flow (although I am also surprised that it it knows the range of outputs for the branchless version).

I don't know how common code like this is, since

```rust
match n {
    n if n > 0 => {}
    0 => {}
    _ => {}
}
```

is much more straightforward than branching on `signum` (I only use `signum` for multiplication), but it's good to know that this PR isn't an unconditional performance win.

**Comment by ecstatic-morse** [comment]
@centril This should get `const-hack`, although there may be a slight performance win from the convoluted version.

**Pull request merged**